File: 1001 - Grid Illumination/solution.cpp
class Solution {
private:
    struct pairHash{
        int operator()(const pair<int, int>& a) const {return a.first^a.second;}
    };
public:
    vector<int> gridIllumination(long long n, vector<vector<int>>& L, vector<vector<int>>& Q) {
        vector<int> res;
        unordered_set<pair<int, int>, pairHash> s;
        unordered_map<int, int> row, col, dia1, dia2;
        for(auto& p : L){
            int x = p[0], y = p[1];
            if(s.insert({x, y}).second)
                row[x]++, col[y]++, dia1[x+y]++, dia2[x-y]++;
        }
        for(auto& p : Q){
            int x = p[0], y = p[1];
            res.push_back(row.count(x)||col.count(y)||dia1.count(x+y)||dia2.count(x-y));
            for(int i = x-1; i <= x+1; i++){
                for(int j = y-1; j <= y+1; j++){
                    if(s.erase({i, j})){
                        if(!--row[i]) row.erase(i);
                        if(!--col[j]) col.erase(j);
                        if(!--dia1[i+j]) dia1.erase(i+j);
                        if(!--dia2[i-j]) dia2.erase(i-j);
                    }
                }
            }
        }
        return res;
    }
};

File: 1001 - Grid Illumination/solution.py
from collections import defaultdict
from itertools import product
class Solution:
    def gridIllumination(self, n: int, L: List[List[int]], Q: List[List[int]]) -> List[int]:
        res, s = [], {(i, j) for i, j in L}
        row, col, dia1, dia2 = defaultdict(int), defaultdict(int), defaultdict(int), defaultdict(int)
        for i, j in s:
            row[i] += 1; col[j] += 1; dia1[i+j] += 1; dia2[i-j] += 1
        for i, j in Q:
            res.append(int(row[i]+col[j]+dia1[i+j]+dia2[i-j] > 0))
            for x, y in product((i-1, i, i+1), (j-1, j, j+1)):
                if (x, y) in s:
                    row[x] -= 1; col[y] -= 1; dia1[x+y] -= 1; dia2[x-y] -= 1
                    s.remove((x, y))
        return res